include/DDpackage.hpp

```diff
@@ -0,0 +1,2047 @@
+        }
+        fp fidelity(const vEdge& x, const vEdge& y) {
+            const auto fid = innerProduct(x, y);
+            return fid.r * fid.r + fid.i * fid.i;
+        }
+
+    private:
+        ComplexValue innerProduct(const vEdge& x, const vEdge& y, Qubit var) {
+            if (x.p == nullptr || y.p == nullptr || x.w.approximatelyZero() || y.w.approximatelyZero()) { // the 0 case
+                return {0.0, 0.0};
+            }
+
+            if (var == 0) {
+                auto c = cn.getTemporary();
+                ComplexNumbers::mul(c, x.w, y.w);
+                return {c.r->value, c.i->value};
+            }
+
+            auto xCopy = x;
+            xCopy.w    = Complex::one;
+            auto yCopy = y;
+            yCopy.w    = Complex::one;
+
+            auto r = vectorInnerProduct.lookup(xCopy, yCopy);
+            if (r.p != nullptr) {
+                auto c = cn.getTemporary(r.w);
+                ComplexNumbers::mul(c, c, x.w);
+                ComplexNumbers::mul(c, c, y.w);
+                return {c.r->val(), c.i->val()};
+            }
+
+            auto w = static_cast<Qubit>(var - 1);
+
+            ComplexValue sum{0.0, 0.0};
+            for (auto i = 0U; i < RADIX; i++) {
+                vEdge e1{};
+                if (!x.isTerminal() && x.p->v == w) {
+                    e1 = x.p->e[i];
+                } else {
+                    e1 = xCopy;
+                }
+                vEdge e2{};
+                if (!y.isTerminal() && y.p->v == w) {
+                    e2   = y.p->e[i];
+                    e2.w = ComplexNumbers::conj(e2.w);
+                } else {
+                    e2 = yCopy;
+                }
+                auto cv = innerProduct(e1, e2, w);
+                sum.r += cv.r;
+                sum.i += cv.i;
+            }
+            r.p = vNode::terminal;
+            r.w = sum;
+
+            vectorInnerProduct.insert(xCopy, yCopy, r);
+            auto c = cn.getTemporary(sum);
+            ComplexNumbers::mul(c, c, x.w);
+            ComplexNumbers::mul(c, c, y.w);
+            return {c.r->val(), c.i->val()};
+        }
+
+        ///
+        /// Kronecker/tensor product
+        ///
+    public:
+        ComputeTable<vEdge, vEdge, vCachedEdge> vectorKronecker{};
+        ComputeTable<mEdge, mEdge, mCachedEdge> matrixKronecker{};
+
+        template<class Node>
+        [[nodiscard]] ComputeTable<Edge<Node>, Edge<Node>, CachedEdge<Node>>& getKroneckerComputeTable();
+
+        template<class Edge>
+        Edge kronecker(const Edge& x, const Edge& y) {
+            auto e = kronecker2(x, y);
+
+            if (e.w != Complex::zero && e.w != Complex::one) {
+                cn.returnToCache(e.w);
+                e.w = cn.lookup(e.w);
+            }
+
+            return e;
+        }
+
+        // extent the DD pointed to by `e` with `h` identities on top and `l` identities at the bottom
+        mEdge extend(const mEdge& e, Qubit h = 0, Qubit l = 0) {
```

Comment:
Does it make sense to have default params which make a call to this method a no-op?

Reply:
I guess not. I removed the default argument on `h`.